libraries/freertos_plus/standard/tls/src/iot_tls.c

```diff
@@ -458,14 +458,18 @@ static int prvInitializeClientCredential( TLSContext_t * pxCtx )
                                                                     sizeof( configPKCS11_DEFAULT_USER_PIN ) - 1 );
     }
 
-    /* Get the handle of the device private key. */
-    xResult = xFindObjectWithLabelAndClass( pxCtx->xP11Session,
-                                            pkcs11configLABEL_DEVICE_PRIVATE_KEY_FOR_TLS,
-                                            CKO_PRIVATE_KEY,
-                                            &pxCtx->xP11PrivateKey );
+    if( CKR_OK == xResult )
+    {
+        /* Get the handle of the device private key. */
+        xResult = xFindObjectWithLabelAndClass( pxCtx->xP11Session,
+                                                pkcs11configLABEL_DEVICE_PRIVATE_KEY_FOR_TLS,
+                                                CKO_PRIVATE_KEY,
+                                                &pxCtx->xP11PrivateKey );
+    }
 
     if( pxCtx->xP11PrivateKey == CK_INVALID_HANDLE )
```

Comment:
If `xResult != CKR_OK` on line 461, is it guaranteed that `pxCtx->xP11PrivateKey` will be set to `CK_INVALID_HANDLE`?  Where was the value of the `xP11PrivateKey` member initialized?

Reply:
Thanks for this - 

The behavior I expect when `xFindObjectWithLabelAndClass` doesn't find an object (but no errors occur) is that it will set xP11PrivateKey to CK_INVALID_OBJECT and return CKR_OK (in the style of PKCS FindObject calls).  I don't know that the standard specifies what happens to the object handle in the event of an error.

An issue that you are highlighting is that if the private key is initialized properly to `CK_INVALID_HANDLE` and we don't even call xFindObjectWithLabelAndClass, the code as it stands would overwrite the previous error code with the 'no private key one' which would be incorrect.   I am going to change the check below to check that CKR_OK is returned AND that the private key is invalid before setting this error code.